The job `x86_64-gnu-llvm-6.0` of your PR [failed](https://dev.azure.com/rust-lang/rust/_build/results?buildId=7438&view=logs&jobId=d0d954b5-f111-5dc4-4d76-03b6c9d0cf7e) ([raw log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/7438/logs/49)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
2019-09-05T15:20:39.1459230Z ##[command]git remote add origin https://github.com/rust-lang/rust
2019-09-05T15:20:39.1475179Z ##[command]git config gc.auto 0
2019-09-05T15:20:39.1477933Z ##[command]git config --get-all http.https://github.com/rust-lang/rust.extraheader
2019-09-05T15:20:39.1485568Z ##[command]git config --get-all http.proxy
2019-09-05T15:20:39.1502083Z ##[command]git -c http.extraheader="AUTHORIZATION: basic ***" fetch --force --tags --prune --progress --no-recurse-submodules --depth=2 origin +refs/heads/*:refs/remotes/origin/* +refs/pull/64177/merge:refs/remotes/pull/64177/merge
---
2019-09-05T15:54:34.7729889Z    Compiling parking_lot_core v0.4.0
2019-09-05T15:54:35.1779425Z    Compiling ena v0.13.0
2019-09-05T15:54:35.6115102Z    Compiling polonius-engine v0.10.0
2019-09-05T15:54:36.7051056Z    Compiling num_cpus v1.8.0
2019-09-05T15:54:37.6407876Z error[E0599]: no method named `serialize_element` found for type `<Self as ser::Serializer>::SerializeSeq` in the current scope
2019-09-05T15:54:37.6408301Z     --> /cargo/registry/src/github.com-1ecc6299db9ec823/serde-1.0.99/src/ser/mod.rs:1275:29
2019-09-05T15:54:37.6408541Z      |
2019-09-05T15:54:37.6408790Z 1275 |             try!(serializer.serialize_element(&item));
2019-09-05T15:54:37.6409237Z      |
2019-09-05T15:54:37.6409639Z      = help: items from traits can only be used if the trait is in scope
2019-09-05T15:54:37.6409639Z      = help: items from traits can only be used if the trait is in scope
2019-09-05T15:54:37.6410276Z      = note: the following trait is implemented but not in scope, perhaps add a `use` for it:
2019-09-05T15:54:37.6410502Z              `use ser::SerializeSeq;`
2019-09-05T15:54:37.6410600Z 
2019-09-05T15:54:37.6578533Z error[E0599]: no method named `serialize_entry` found for type `<Self as ser::Serializer>::SerializeMap` in the current scope
2019-09-05T15:54:37.6578927Z     --> /cargo/registry/src/github.com-1ecc6299db9ec823/serde-1.0.99/src/ser/mod.rs:1315:29
2019-09-05T15:54:37.6579188Z      |
2019-09-05T15:54:37.6579481Z 1315 |             try!(serializer.serialize_entry(&key, &value));
2019-09-05T15:54:37.6580017Z      |
2019-09-05T15:54:37.6580309Z      = help: items from traits can only be used if the trait is in scope
2019-09-05T15:54:37.6580309Z      = help: items from traits can only be used if the trait is in scope
2019-09-05T15:54:37.6580623Z      = note: the following trait is implemented but not in scope, perhaps add a `use` for it:
2019-09-05T15:54:37.6580869Z              `use ser::SerializeMap;`
2019-09-05T15:54:38.3507533Z    Compiling jobserver v0.1.16
2019-09-05T15:54:38.7847341Z error: aborting due to 2 previous errors
2019-09-05T15:54:38.7853180Z 
2019-09-05T15:54:38.7859684Z For more information about this error, try `rustc --explain E0599`.
---
2019-09-05T15:54:40.5035858Z == clock drift check ==
2019-09-05T15:54:40.5055612Z   local time: Thu Sep  5 15:54:40 UTC 2019
2019-09-05T15:54:40.6630938Z   network time: Thu, 05 Sep 2019 15:54:40 GMT
2019-09-05T15:54:40.6635965Z == end clock drift check ==
2019-09-05T15:54:42.0522237Z ##[error]Bash exited with code '1'.
2019-09-05T15:54:42.0562440Z ##[section]Starting: Checkout
2019-09-05T15:54:42.0565032Z ==============================================================================
2019-09-05T15:54:42.0565090Z Task         : Get sources
2019-09-05T15:54:42.0565157Z Description  : Get sources from a repository. Supports Git, TfsVC, and SVN repositories.
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

--- a/src/librustc_resolve/build_reduced_graph.rs
+++ b/src/librustc_resolve/build_reduced_graph.rs
@@ -728,9 +728,10 @@ impl<'a, 'b> BuildReducedGraphVisitor<'a, 'b> {
                                              expansion,
                                              item.span);
                 self.r.define(parent, ident, TypeNS, (module, vis, sp, expansion));
+                self.parent_scope.module = module;
 
                 for variant in &(*enum_definition).variants {
-                    self.build_reduced_graph_for_variant(variant, module, vis);
+                    self.build_reduced_graph_for_variant(variant, vis);
                 }
             }
 
@@ -818,10 +819,8 @@ impl<'a, 'b> BuildReducedGraphVisitor<'a, 'b> {
 
     // Constructs the reduced graph for one variant. Variants exist in the
     // type and value namespaces.
-    fn build_reduced_graph_for_variant(&mut self,
-                                       variant: &Variant,
-                                       parent: Module<'a>,
-                                       vis: ty::Visibility) {
+    fn build_reduced_graph_for_variant(&mut self, variant: &Variant, vis: ty::Visibility) {
+        let parent = self.parent_scope.module;
         let expn_id = self.parent_scope.expansion;
         let ident = variant.ident;
 
@@ -1253,9 +1252,7 @@ impl<'a, 'b> Visitor<'b> for BuildReducedGraphVisitor<'a, 'b> {
         let expansion = self.parent_scope.expansion;
         self.r.define(parent, item.ident, ns, (res, vis, item.span, expansion));
 
-        self.parent_scope.module = parent.parent.unwrap(); // nearest normal ancestor
         visit::walk_trait_item(self, item);
-        self.parent_scope.module = parent;
     }
 
     fn visit_token(&mut self, t: Token) {

--- a/src/librustc_resolve/lib.rs
+++ b/src/librustc_resolve/lib.rs
@@ -538,7 +538,11 @@ impl<'a> ModuleData<'a> {
     }
 
     fn nearest_item_scope(&'a self) -> Module<'a> {
-        if self.is_trait() { self.parent.unwrap() } else { self }
+        match self.kind {
+            ModuleKind::Def(DefKind::Enum, ..) | ModuleKind::Def(DefKind::Trait, ..) =>
+                self.parent.expect("enum or trait module without a parent"),
+            _ => self,
+        }
     }
 
     fn is_ancestor_of(&self, mut other: &Self) -> bool {

--- a/src/librustc_resolve/lib.rs
+++ b/src/librustc_resolve/lib.rs
@@ -1644,7 +1644,7 @@ impl<'a> Resolver<'a> {
         }
 
         if let ModuleKind::Block(..) = module.kind {
-            return Some(module.parent.unwrap());
+            return Some(module.parent.unwrap().nearest_item_scope());
         }
 
         None

--- a/src/test/ui/resolve/block-with-trait-parent.rs
+++ b/src/test/ui/resolve/block-with-trait-parent.rs
@@ -1,0 +1,14 @@
+// check-pass
+
+trait Trait {
+    fn method(&self) {
+        // Items inside a block turn it into a module internally.
+        struct S;
+        impl Trait for S {}
+
+        // OK, `Trait` is in scope here from method resolution point of view.
+        S.method();
+    }
+}
+
+fn main() {}

